core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp

```diff
@@ -0,0 +1,85 @@
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_PARALLEL_FOR_RANGE_HPP
+#define KOKKOS_OPENACC_PARALLEL_FOR_RANGE_HPP
+
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp>
+#include <Kokkos_Parallel.hpp>
+
+template <class Functor, class... Traits>
+class Kokkos::Impl::ParallelFor<Functor, Kokkos::RangePolicy<Traits...>,
+                                Kokkos::Experimental::OpenACC> {
+  using Policy = Kokkos::RangePolicy<Traits...>;
+  Kokkos::Experimental::Impl::FunctorAdapter<Functor, Policy> m_functor;
+  Policy m_policy;
+
+ public:
+  ParallelFor(Functor const& functor, Policy const& policy)
+      : m_functor(functor), m_policy(policy) {}
+
+  void execute() const {
+    auto const begin = m_policy.begin();
+    auto const end   = m_policy.end();
+
+    if (end <= begin) {
+      Kokkos::Impl::throw_runtime_exception(std::string(
+          "Kokkos::Impl::ParallelFor< OpenACC > can not be executed with "
+          "a range <= 0."));
+    }
+
+    int const async_arg = m_policy.space().acc_async_queue();
+    // avoid implicit capture of *this which would yield a memory access
+    // violation when executing on the device
+    auto const& a_functor(m_functor);
+
+#pragma acc parallel loop gang vector copyin(a_functor) async(async_arg)
```

Comment:
Why "gang vector"?

Reply:
OpenACC provides three levels of parallelism (gang/worker/vector), but most of existing GPUs (NVIDIA GPU and AMD GPU) provides only two levels of parallelism, and thus mapping OpenACC three level parallelism to the GPU's two level parallelism has been issues for a long time (and each OpenACC compiler (e.g., NVHPC and GCC) also changed their mapping strategies multiple times). For now, just using both gang and vector seems to be the best choice at least for both NVHPC and GCC.

Reply:
`acc parallel loop …` may work well for NVHPC, but GCC seems to be less efficient in handling auto parallelization and thus may not work well.
A bigger issue is that an OpenACC compiler may not be able to figure out what kind of parallelisms are included in the functor, in which case, the compiler will not be able to find which parallelisms can be allowed to this loop. 
Therefore, it will be better to keep `loop gang vector` as it is.

Reply:
It can look weird using the first (gang) and last (vector) OpenACC parallelism level together, but as Seyong commented, using gang and vector works better

Reply:
do a ifdef based on compiler: for nvhpc use `acc parallel loop`, and for gcc `acc parallel loop gang vector`

Reply:
and then lets complain to the vendors. So generally I want to see at least a comment what we think we should ideally use, vs what we need to use because the compiler is dumb

Reply:
`acc parallel loop gang vector` is the best choice for NVHPC too, and even the NVHPC compiler may have the same limit on handling functors; if the compiler cannot figure out available parallelisms in the functor, `acc parallel loop` will not work.